### PR TITLE
Add logical_and and logical_or

### PR DESCRIPTION
**Pull request by xuhdev**
Superseding #24379 as type promotion has been implemented.

Close #24379

### CODE REVIEW

**Comment by xuhdev** [comment]
@ifedan @gchanan Interested in review? :) This should be pretty similar to logical_xor #27248

**Comment by gchanan** [comment]
@nairbv can you review?

File: aten/src/ATen/native/quantized/README.md [hunk 1]
```diff
@@ -7,24 +7,24 @@ This document serves as an entry point for quantized kernel implementation.
 ## Implementing native quantized ops
 
 The new quantized ops are almost always located under the `ATen/native/quantized/cpu` folder.
-For the sake of an example, let us implement an element-wise quantized logical AND operation under `ATen/native/quantized/cpu/qand.cpp`.
+For the sake of an example, let us implement an element-wise quantized logical XAND operation under `ATen/native/quantized/cpu/qxand.cpp`.
```
Comment (nairbv) [inline aten/src/ATen/native/quantized/README.md]: what is "xand"? is this related to the rest of the PR?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -602,6 +602,38 @@
     CUDA: logical_xor_out
   supports_named_tensor: True
 
+- func: logical_and(Tensor self, Tensor other) -> Tensor
+  use_c10_dispatcher: unboxed_only
+  variants: function, method
+  supports_named_tensor: True
```
Comment (nairbv) [inline aten/src/ATen/native/native_functions.yaml]: since these support naming, we should have tests in test_namedtensor.py like we have for test_logical_xor.

File: aten/src/ATen/native/quantized/README.md (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: This is to replace an example in quantized outmoded by this patch. AND has already been implemented, so I replaced it with XAND.

Comment (nairbv) [reply aten/src/ATen/native/quantized/README.md]: does 'xand' mean anything or it's just a dummy op-name for an example? If it's just an example, maybe we can name it something clearer?

Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: Yes, [xand](https://en.wiktionary.org/wiki/XAND) has its own meaning

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -2667,6 +2667,34 @@ def merge_dicts(*dicts):
...
+
+Args:
+    {input}
+    other (Tensor): the tensor to compute AND with
+    {out}
+
+Example::
+
+    >>> torch.logical_and(torch.tensor([True, False, True]), torch.tensor([True, False, False]))
+    tensor([ False, False,  True])
```
Comment (nairbv) [inline torch/_torch_docs.py]: Extra space chars make me think this example isn't pasted from a python prompt -- I think the last value should be `False` for `and` right? True and False for index `2`?

File: aten/src/ATen/native/quantized/README.md (hunk 1 above)
Comment (nairbv) [reply aten/src/ATen/native/quantized/README.md]: that seems a bit confusing to me, but I guess that's OK

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -2690,6 +2718,34 @@ def merge_dicts(*dicts):
...
+
+Args:
+    {input}
+    other (Tensor): the tensor to compute OR with
+    {out}
+
+Example::
+
+    >>> torch.logical_or(torch.tensor([True, False, True]), torch.tensor([True, False, False]))
+    tensor([ False, False,  True])
```
Comment (nairbv) [inline torch/_torch_docs.py]: True or True should be True?

File: torch/_torch_docs.py (hunk 3 above)
Comment (xuhdev) [reply torch/_torch_docs.py]: This is probably my negligence in copying over the example from `logical_xor`. I've corrected this example now

File: aten/src/ATen/native/quantized/README.md (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: I've added a hyperlink to an explanation of xand

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -2690,6 +2718,34 @@ def merge_dicts(*dicts):
...
+    {out}
+
+Example::
+
+    >>> torch.logical_or(torch.tensor([True, False, True]), torch.tensor([True, False, False]))
+    tensor([ True, False,  True])
+    >>> a = torch.tensor([0, 1, 10, 0], dtype=torch.int8)
+    >>> b = torch.tensor([4, 0, 1, 0], dtype=torch.int8)
+    >>> torch.logical_or(a, b)
+    tensor([ True,  True,  True, False])
```
Comment (nairbv) [inline torch/_torch_docs.py]: let's fix the extra spaces too so this actually looks like a direct paste

Comment (xuhdev) [reply torch/_torch_docs.py]: I think this is pretty print as it is what was printed from my Python shell.

Comment (nairbv) [reply torch/_torch_docs.py]: ah I see, sorry, yeah I get the same when printing a bool tensor.

**Review by nairbv — APPROVED** [review]

File: aten/src/ATen/native/BinaryOps.cpp [hunk 6]
```diff
@@ -308,5 +310,18 @@ Tensor& logical_xor_out(Tensor& result, const Tensor& self, Scalar other) { retu
 Tensor logical_xor(const Tensor& self, Scalar other) { return comparison_op(self, other, logical_xor_stub); }
 Tensor& logical_xor_(Tensor& self, Scalar other) { return comparison_op_(self, other, logical_xor_stub); }
 
+Tensor& logical_and_out(Tensor& result, const Tensor& self, const Tensor& other) { return comparison_op_out(result, self, other, logical_and_stub); }
```
Comment (ailzhang) [inline aten/src/ATen/native/BinaryOps.cpp]: Hi @xuhdev I think I found the root cause why these `logical_*` ops are not working for XLA. 

The problem is in the implementation of `comparison_op` that they should be implemented similar to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UnaryOps.cpp#L52. The current implementation directly call into TensorIterator through stubs which didn't give a chance for device dispatch to get in.
(XLA will be called through logical_xor_out :)

**Review by ailzhang — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/BinaryOps.cpp (hunk 6 above)
Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: Thanks. I think you actually referred to calling `at::` instead of `native::`? I've pushed a change for testing (eventually we should move the change out to a new PR). Interestingly, the comparison operators (ge, le, ...) are not skipped in the XLA CI.

Comment (ailzhang) [reply aten/src/ATen/native/BinaryOps.cpp]: @xuhdev `at::comparison_op` won't work because you only have it in `at::native` namespace. But if you pass in the function `out_impl` instead of stub it should work. 
`template <typename OutImpl>
static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {`

Comment (ailzhang) [reply aten/src/ATen/native/BinaryOps.cpp]: Yea in `ge`s case it's different sorry it's indeed confusing. 
So `ge` is a non-composable op that XLA is called at `ge` level. But in `logical_*` XLA is expected to be called at `logical_*_out` level. We can change that but need to be a bit careful. 
I was going to write a wiki to explain this better for our OSS contributors, but a quick reference: https://gist.github.com/ailzhang/905aa41a5be8c5315ba0e3d8f0d0e326 contains all currently non-composable ops in Pytorch codebase. which means XLA will be called at these ops.

Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: Thanks, this info is indeed helpful. Is there going to be a plan to make all ops either composable or non-composable? I'm just trying to understand what the ultimate design is :)

Comment (ailzhang) [reply aten/src/ATen/native/BinaryOps.cpp]: No we'll keep both I think since for non-composable ops autograd is handling them to save a lot of manual work, and that's indeed helpful. 
But we'll definitely look into ways to make this easier to tell for OSS contributors. Another easy way to judge is to look at `build/aten/src/ATen/Declarations.yaml` and check where `abstract=true` for the functions. If you have ideas what could help you understand, we're happy to hear your thoughts as well!

**Comment by xuhdev** [comment]
Blocked by #29591

File: docs/source/tensors.rst [hunk 7]
```diff
@@ -327,8 +327,12 @@ view of a storage and defines numeric operations on it.
    .. automethod:: log2_
    .. automethod:: log_normal_
    .. automethod:: logsumexp
+   .. automethod:: logical_and
```
Comment (ailzhang) [inline docs/source/tensors.rst]: I'm curious do we not have public apis for `logical_and` before?

**Review by ailzhang — CHANGES_REQUESTED** [review]
@xuhdev Thanks! 
The test failures in normal pytorch tests seem valid. 
cc: @jerryzh168 to review the quantization related changes.

**Comment by xuhdev** [comment]
Yeah this is a real failure. As a note, the following situation fails, possibly related to promotion:

    >>> a= torch.tensor([10, 0, 1, 0], dtype=torch.uint8)
    >>> b= torch.tensor([1, 0, 0, 10], dtype=torch.float16)
    >>> a.logical_and(b)
    tensor([False, False, False, False])

File: docs/source/tensors.rst (hunk 7 above)
Comment (xuhdev) [reply docs/source/tensors.rst]: I don't see any (while there are bitwise ones)

**Comment by xuhdev** [comment]
Turned out I mistyped in the CPU kernel functions. Updated

File: aten/src/ATen/native/quantized/README.md [hunk 8]
```diff
@@ -6,25 +6,26 @@ This document serves as an entry point for quantized kernel implementation.
 
 ## Implementing native quantized ops
 
-The new quantized ops are almost always located under the `ATen/native/quantized/cpu` folder.
-For the sake of an example, let us implement an element-wise quantized logical AND operation under `ATen/native/quantized/cpu/qand.cpp`.
+The new quantized ops are almost always located under the `ATen/native/quantized/cpu` folder. For
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/README.md]: why do you change `README` of implementing quantized kernels?

Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: Because the old example that implement `and` is outmoded by this PR. I replaced it with an example operator that does not exist after this PR

File: aten/src/ATen/native/quantized/README.md [hunk 9]
```diff
@@ -6,25 +6,26 @@ This document serves as an entry point for quantized kernel implementation.
...
   double scale = qa.q_scale();
   long zero_point qa.q_zero_point();
 
   auto iter = TensorIterator::binary_op(qc, qa, qb);
 
-  AT_DISPATCH_QINT_TYPES(qa.scalar_type(), "quantized_and", [&]() {
+  AT_DISPATCH_QINT_TYPES(qa.scalar_type(), "quantized_xand", [&]() {
     Tensor qc = at::_empty_affine_quantized(
         qa.sizes(), at::device(kCPU).dtype(SCALAR_TYPE), scale, zero_point);
     cpu_kernel(iter, [&](scalar_t a_value, scalar_t b_value) -> scalar_t {
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/README.md]: I didn't see change of implementation from `and` to `xand`, do you need to change the code below as well?

Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: I believe the implementation is abstractly summarized as:

```c++
  Tensor quantized_xand(Tensor qa, Tensor qb) {
    // The awesome op implementation...
    return qc;
  }
```

**Comment by xuhdev** [comment]
@ailzhang All tests have passed now

**Review by ailzhang — APPROVED** [review]
Thanks!

**Comment by xuhdev** [comment]
A friendly reminder: Is this PR ready to merge? :)

**Comment by ailzhang** [comment]
@xuhdev Thanks for the reminder! Just realized the land failed :( I'll try again in the afternoon.

**Comment by xuhdev** [comment]
Thanks!

**Comment by zou3519** [comment]
Master CI is failing. In particular, the test for `test/test_torch.py::TestTorchDeviceTypeCUDA::test_logical_and_cuda` is failing and it was added in this PR: https://app.circleci.com/jobs/github/pytorch/pytorch/3730054.

I have confirmed locally that this test fails when run by itself (the other CUDA test failures in the log are because the CUDA device gets messed up after one failure). I am reverting this PR for now, please feel free to resubmit it later.
